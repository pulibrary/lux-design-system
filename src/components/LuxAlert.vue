<template>
  <transition name="fade">
    <div
      v-if="show"
      class="lux-alert"
      :class="{
        'lux-alert-dismissible': dismissible,
        'lux-alert-success': isSuccess,
        'lux-alert-warning': isWarning,
        'lux-alert-error': isError,
        'lux-alert-info': isInfo,
        'lux-alert-fullscreen': isFullScreen,
      }"
      role="alert"
    >
      <span>
        <slot>{{ alertMessage }}</slot>
      </span>
      <button
        v-if="dismissible"
        type="button"
        class="lux-close"
        data-dismiss="alert"
        aria-label="Close"
        @click="hideAlert()"
      >
        <span aria-hidden="true">×</span>
      </button>
    </div>
  </transition>
</template>

<script>
/**
 * Alerts are used to provide timely information to a user in response to some event.
 */
export default {
  name: "LuxAlert",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  metaInfo: {
    title: "Global alert message to user.",
    htmlAttrs: {
      lang: "en",
    },
  },
  data: function () {
    return {
      alertMessage: "Something happened, but we're not sure what.",
      show: true,
    }
  },
  computed: {
    isSuccess() {
      return this.status === "success" ? true : false
    },
    isInfo() {
      return this.status === "info" ? true : false
    },
    isWarning() {
      return this.status === "warning" ? true : false
    },
    isError() {
      return this.status === "error" ? true : false
    },
    isFullScreen() {
      return this.type === "alert" ? true : false
    },
  },
  props: {
    /**
     * The intent of the message. Valid options: `alert, indicator`. Alerts are full page and indicators are “inline”.
     */
    type: {
      type: String,
      default: "indicator",
      validator: value => {
        return value.match(/(alert|indicator)/)
      },
    },
    /**
     * Severity of the message. Valid options: `info, warning, success, error`
     */
    status: {
      type: String,
      default: "info",
      validator: value => {
        return value.match(/(info|warning|success|error)/)
      },
    },
    /**
     * Automatically hides the notification after 2 seconds.
     */
    autoclear: {
      type: Boolean,
      default: false,
    },
    /**
     * User can manually hide the notification.
     */
    dismissible: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    hideAlert() {
      this.show = false
    },
  },
  mounted() {
    if (this.autoclear) {
      setTimeout(() => {
        this.show = false
      }, 2000)
    }
  },
}
</script>

<style lang="scss">
@import "../assets/styles/variables.css";
@import "../assets/styles/mixins.scss";

.lux-alert-fullscreen {
  position: fixed;
  top: 5vh;
  left: 50%; /* move the left edge to the center … */
  margin-left: -25vw !important; /* … and move it to the left half the box’ width. */
  z-index: 9999;
  width: 50vw;
}

.lux-alert {
  @include reset;
  font-family: var(--font-family-text);
  padding: 0.75rem 1.25rem;
  margin-bottom: 1rem;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  position: relative;
}
.lux-alert-success {
  background-color: #dff0d8;
  border-color: #d0e9c6;
  color: #3c763d;
}
.lux-alert-error {
  background-color: #f2dede;
  border-color: #ebcccc;
  color: #a94442;
}
.lux-alert-warning {
  background-color: #fcf8e3;
  border-color: #faf2cc;
  color: #8a6d3b;
}
.lux-alert-info {
  background-color: #d9edf7;
  border-color: #bcdff1;
  color: #31708f;
}

.lux-alert-dismissible > span {
  width: calc(100% - 25px);
  display: inline-block;
}

.lux-alert-dismissible .lux-close {
  position: absolute;
  top: -4px;
  right: -5px;
  padding: 0.75rem 1.25rem;
  color: inherit;
}

button.lux-close {
  padding: 0;
  cursor: pointer;
  background: 0 0;
  border: 0;
  -webkit-appearance: none;
}

.lux-close {
  float: right;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1;
  color: #000;
  text-shadow: 0 1px 0 #fff;
  opacity: 0.5;
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}
.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>

<docs>
  ```jsx
    <div>
      <lux-alert status="warning" autoclear>How to disappear completely...</lux-alert>
      <lux-alert status="error"></lux-alert>
      <lux-alert status="success">Like a boss!</lux-alert>
      <lux-alert status="info" dismissible>Here's some dismissible info for you.</lux-alert>
    </div>
  ```
</docs>
