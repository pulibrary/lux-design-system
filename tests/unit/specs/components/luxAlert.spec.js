import { nextTick } from "vue"
import { mount } from "@vue/test-utils"
import LuxAlert from "@/components/LuxAlert.vue"

const transitionStub = () => ({
  render: function (h) {
    return this.$options._renderChildren
  },
})

describe("LuxAlert.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(LuxAlert, {
      slots: {
        default: "Here's some info for you.",
      },
      stubs: {
        transition: transitionStub(),
      },
    })
  })

  it("should have an alert message", () => {
    const el = wrapper.find(".lux-alert")
    expect(el.text()).toBe("Here's some info for you.")
  })

  it("should have a button when dismissible", async () => {
    wrapper.setProps({ dismissible: true })
    await nextTick()
    const button = wrapper.find("button")
    expect(wrapper.vm.dismissible).toBe(true)
    expect(button.exists()).toBe(true)
  })

  it("should be dismissible on click", async () => {
    wrapper.setProps({ dismissible: true })
    await nextTick()
    expect(wrapper.findComponent({ name: "LuxAlert" })).toBeTruthy
    const button = wrapper.find("button")
    button.trigger("click")
    await nextTick()
    expect(wrapper.findComponent({ name: "LuxAlert" })).toBeFalsy
  })

  it("should be accessible", () => {
    const el = wrapper.find(".lux-alert")
    expect(el.attributes().role).toBe("alert")
  })

  it("should default to info type", () => {
    expect(wrapper.vm.isInfo).toBe(true)
  })

  it("should hide alert when hideAlert() is called", () => {
    expect(wrapper.vm.show).toBe(true)
    wrapper.vm.hideAlert()
    expect(wrapper.vm.show).toBe(false)
  })

  it("should hide after 2 seconds when autoclear is true", async () => {
    wrapper.setProps({ autoclear: true })
    await nextTick()
    const el = wrapper.find(".lux-alert")
    expect(wrapper.vm.show).toBe(true)
    // wait 3 seconds and see if it's hidden
    setTimeout(() => {
      expect(wrapper.vm.show).toBe(false)
    }, 3000)
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
