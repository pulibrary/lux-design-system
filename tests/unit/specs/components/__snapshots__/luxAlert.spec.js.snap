// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxAlert.vue has the expected html structure 1`] = `
<transition-stub
  appear="false"
  css="true"
  name="fade"
  persisted="false"
>
  <div
    class="lux-alert lux-alert-info"
    role="alert"
  >
    <span>
      <!-- @slot the alert text -->
      
      Here's some info for you.
      
    </span>
    <!--v-if-->
  </div>
</transition-stub>
`;
